typescript/sdk/src/token/EvmERC20WarpRouteReader.ts

```diff
@@ -44,41 +67,65 @@ export class EvmERC20WarpRouteReader {
    */
   async deriveWarpRouteConfig(
     address: Address,
-  ): Promise<DerivedERC20WarpRouteConfig> {
+  ): Promise<DerivedTokenRouterConfig> {
+    // Derive the config type
+    const type = await this.deriveTokenType(address);
     const fetchedBaseMetadata = await this.fetchBaseMetadata(address);
-    const fetchedTokenMetadata = await this.fetchTokenMetadata(
-      fetchedBaseMetadata.token,
-    );
+    const fetchedTokenMetadata = await this.fetchTokenMetadata(type, address);
 
-    const results: DerivedERC20WarpRouteConfig = {
+    return {
       ...fetchedBaseMetadata,
       ...fetchedTokenMetadata,
+      type,
+    } as DerivedTokenRouterConfig;
+  }
+
+  /**
+   * Derives the token type for a given Warp Route address using specific methods
+   *
+   * @param address - The Warp Route address to derive the token type for.
+   * @returns The derived token type, which can be one of: collateralVault, collateral, native, or synthetic.
+   */
+  async deriveTokenType(address: Address): Promise<DerivedTokenType> {
+    const contractTypes: Record<
+      Exclude<DerivedTokenType, 'native'>, // native is excluded because it's the default return type
+      { factory: any; method: string }
+    > = {
+      collateral: {
+        factory: HypERC20Collateral__factory,
+        method: 'wrappedToken',
+      },
+      collateralVault: {
+        factory: HypERC20CollateralVaultDeposit__factory,
+        method: 'vault',
+      },
+      synthetic: {
+        factory: HypERC20__factory,
+        method: 'decimals',
+      },
     };
 
-    if (
-      fetchedBaseMetadata.interchainSecurityModule !==
-      ethers.constants.AddressZero
-    ) {
-      results.interchainSecurityModule =
-        await this.evmIsmReader.deriveIsmConfig(
-          fetchedBaseMetadata.interchainSecurityModule,
+    for (const [type, { factory, method }] of Object.entries(contractTypes)) {
+      try {
+        const warpRoute = factory.connect(address, this.provider);
+        await warpRoute[method]();
+        return type as DerivedTokenType;
+      } catch (e) {
+        this.logger.debug(
+          `Error accessing token specific property, implying this is not a ${type} token. Defaulting to ${TokenType.native}.`,
```

Comment:
this defaulting comment is only true if you fail for all three methods, right? would suggest adding the log line outside of the for loop if so

Reply:
This comment should be displayed each time it fails, before the next iteration. The second part is confusing, and I will move it outside. also updated comment